code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -348,3 +350,86 @@ pub fn compute_deposit_lp(
 		Ok((overwritten_quote_amount, lp_to_mint))
 	}
 }
+
+/// Computes the decimal value of a a `PerThing` by taking the deconstructed parts of a `PerThing`
+/// and dividing them by `PerThing::one()`.
+///
+/// # Example
+/// ```rust,ignore
+/// let per_thing = PerMill::from_percent(10);
+/// assert_eq!(decimal_from_per_thing(per_thing), Decimal::new(10, 2));
+/// ```
+fn decimal_from_per_thing<T: PerThing>(per_thing: T) -> Result<Decimal, ArithmeticError> {
+	let numerator =
+		Decimal::from_u128(per_thing.deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+	let denominator =
+		Decimal::from_u128(T::one().deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+
+	numerator.safe_div(&denominator)
+}
+
+/// Computes the LP to mint on first deposit.
+///
+/// If `Ok`, returns a tuple containing `(lp_to_mint, fee)`.
+///
+/// Fees are currently always 0. Fees are normally charged to avoid fee-less swaps by adding and
+/// removing liquidity. With the initial deposit, these chances are so low that it is safe to
+/// process without a fee.
+///
+/// Paramaters:
+/// * `lp_token_details` - Vec of tuples containing `(token_deposit, token_balance, token_weight)`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 6
+fn compute_first_deposit_lp_<T: PerThing>(
+	pool_assets: Vec<(u128, u128, T)>,
```

Comment:
```suggestion
	pool_assets: &[(u128, u128, T)],
```

No need to take ownership of the vec here; also consider a struct instead of a tuple

Reply:
024c8bb90c1929efcac8312609c866911c59b5df